### PR TITLE
[Locator] Make it better and even more incredible.

### PR DESCRIPTION
**Pull request by kobolog**
Make locator thread-safe since the Unicorn Cluster will issue `link_node()` and `drop_node()` from random ZK client threads, and also drop useless crap.

### CODE REVIEW

**Comment by kobolog** [comment]
@3Hren @antmat: WTB review!

**Comment by kobolog** [comment]
So, the `api::client<T>` is not thread-safe, but Locator makes sure it's synchronized across the place. Everything else should be thread-safe now.

File: include/cocaine/api/connect.hpp [hunk 1]
```diff
@@ -72,8 +66,11 @@ template<class Tag>
 class client:
     public details::basic_client_t
 {
+    template<class Event, bool = std::is_same<typename Event::tag, Tag>::value>
+    struct traits;
+
     template<class Event>
-    struct traits {
+    struct traits<Event, true> {
```
Comment (antmat) [inline include/cocaine/api/connect.hpp]: What's the benefit of weird compilation error on static assert with pretty looking message?

Comment (kobolog) [inline include/cocaine/api/connect.hpp]: Static assert error output looks even weirder that this on GCC 4.6.

**Pull request merged**